xds/internal/xdsclient/transport/ads/ads_stream.go

```diff
@@ -0,0 +1,780 @@
+			if err := s.sendMessageLocked(stream, names, typ.TypeURL(), state.version, state.nonce, nil); err != nil {
+				return err
+			}
+			s.startWatchTimersLocked(typ, names)
+		default:
+			// No buffered request.
+			continue
+		}
+	}
+	return nil
+}
+
+// sendMessageLocked sends a discovery request to the server, populating the
+// different fields of the message with the given parameters. Returns a non-nil
+// error if the request could not be sent.
+//
+// Caller needs to hold c.mu.
+func (s *Stream) sendMessageLocked(stream transport.StreamingCall, names []string, url, version, nonce string, nackErr error) error {
+	req := &v3discoverypb.DiscoveryRequest{
+		ResourceNames: names,
+		TypeUrl:       url,
+		VersionInfo:   version,
+		ResponseNonce: nonce,
+	}
+
+	// The xDS protocol only requires that we send the node proto in the first
+	// discovery request on every stream. Sending the node proto in every
+	// request wastes CPU resources on the client and the server.
+	if s.firstRequest {
+		req.Node = s.nodeProto
+	}
+
+	// If this is a NACK, populate the error_detail field.
+	if nackErr != nil {
+		req.ErrorDetail = &statuspb.Status{
+			Code: int32(codes.InvalidArgument), Message: nackErr.Error(),
+		}
+	}
+
+	if err := stream.Send(req); err != nil {
+		s.logger.Warningf("Sending ADS request for type %q, resources: %v, version: %q, nonce: %q failed: %v", url, names, version, nonce, err)
+		return err
+	}
+	s.firstRequest = false
+
+	if s.logger.V(perRPCVerbosityLevel) {
+		s.logger.Infof("ADS request sent: %v", pretty.ToJSON(req))
+	} else {
+		if s.logger.V(2) {
+			s.logger.Warningf("ADS request sent for type %q, resources: %v, version: %q, nonce: %q", url, names, version, nonce)
+		}
+	}
+	return nil
+}
+
+// Return value indicates if at least one message was received from the server.
+func (s *Stream) recv(ctx context.Context, stream transport.StreamingCall) bool {
+	msgReceived := false
+	for {
+		// Wait for ADS stream level flow control to be available.
+		if !s.fc.wait(ctx) {
+			if s.logger.V(2) {
+				s.logger.Infof("ADS stream context canceled")
+			}
+			return msgReceived
+		}
+
+		// Send out a request if anything was buffered while we were waiting for
+		// local processing of the previous response to complete.
+		s.sendBuffered(stream)
+
+		resources, url, version, nonce, err := s.recvMessage(stream)
+		if err != nil {
+			// Note that we do not consider it an error if the ADS stream was closed
+			// after having received a response on the stream. This is because there
+			// are legitimate reasons why the server may need to close the stream during
+			// normal operations, such as needing to rebalance load or the underlying
+			// connection hitting its max connection age limit.
+			// (see [gRFC A9](https://github.com/grpc/proposal/blob/master/A9-server-side-conn-mgt.md)).
+			if msgReceived {
+				err = xdsresource.NewErrorf(xdsresource.ErrTypeStreamFailedAfterRecv, err.Error())
+			}
+			s.onError(err, msgReceived)
```

Comment:
onError already wraps in ErrTypeStreamFailedAfterRecv. This would wrap it twice. I don't get the docstring "we do not consider it an error if the ADS stream was closed after having received a response on the stream". We're still passing it to the channel no matter the type. Would the channel ignore it? If so please specify that in docstring.